Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+{
+    VERIFY(!insn.modrm().is_register());
+    auto f = fpu_get(0);
+    set_c1(0);
+    auto int32 = fpu_round_checked<i32>(f);
+    // FIXME: Respect shadow values
+    insn.modrm().write32(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u32>(int32)));
+}
+
+void SoftFPU::FISTP_RM16(const X86::Instruction& insn)
+{
+    FIST_RM16(insn);
+    fpu_pop();
+}
+void SoftFPU::FISTP_RM32(const X86::Instruction& insn)
+{
+    FIST_RM32(insn);
+    fpu_pop();
+}
+void SoftFPU::FISTP_RM64(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto f = fpu_pop();
+    set_c1(0);
+    auto i64 = fpu_round_checked<int64_t>(f);
+    // FIXME: Respect shadow values
+    insn.modrm().write64(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u64>(i64)));
+}
+
+void SoftFPU::FISTTP_RM16(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    set_c1(0);
+    i16 value = static_cast<i16>(fpu_pop());
+    // FIXME: Respect shadow values
+    insn.modrm().write16(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u16>(value)));
+}
+void SoftFPU::FISTTP_RM32(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    i32 value = static_cast<i32>(fpu_pop());
+    set_c1(0);
+    // FIXME: Respect shadow values
+    insn.modrm().write32(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u32>(value)));
+}
+void SoftFPU::FISTTP_RM64(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    set_c1(0);
+    i64 value = static_cast<i64>(fpu_pop());
+    // FIXME: Respect shadow values
+    insn.modrm().write64(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u64>(value)));
+}
+
+void SoftFPU::FBLD_M80(const X86::Instruction&) { TODO_INSN(); }
+void SoftFPU::FBSTP_M80(const X86::Instruction&) { TODO_INSN(); }
+
+void SoftFPU::FXCH(const X86::Instruction& insn)
+{
+    // FIXME: implicit argument `D9 C9` -> st[0] <-> st[1]?
+    VERIFY(insn.modrm().is_register());
+    set_c1(0);
+    auto tmp = fpu_get(0);
+    fpu_set(0, fpu_get(insn.modrm().register_index()));
+    fpu_set(insn.modrm().register_index(), tmp);
+}
+
+void SoftFPU::FCMOVE(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    if (m_cpu.zf())
+        fpu_set(0, fpu_get(insn.modrm().rm()));
+}
+void SoftFPU::FCMOVNE(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    if (!m_cpu.zf())
+        fpu_set(0, fpu_get((insn.modrm().reg_fpu())));
+}
+
+void SoftFPU::FCMOVB(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    if (m_cpu.cf())
+        fpu_set(0, fpu_get(insn.modrm().rm()));
+}
+void SoftFPU::FCMOVNB(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m32int = (i32)insn.modrm().read32(m_cpu, insn).value();
+    // FIXME: Respect shadow values
+    fpu_push((long double)m32int);
+}
+void SoftFPU::FCMOVBE(const X86::Instruction& insn)
+{
+    if (m_cpu.cf() | m_cpu.zf())
```

Comment:
Would prefer || or `or` for binary or.

Reply:
Also, reason to prefer FCMOVBE instead of FCMOVA ?

Reply:
FCMOVA?

Reply:
CMOVcc has postfixes for B (Below) and BE (Below-Equal), with NBE being an alias for A (not below-equal == above) and NB being an alias for AE (Not-Below == Above-Equal). Apparently that doesn't translate to FCMOVcc... 

Comment should've been on FCMOVNBE, which I would read more readily as FCMOVA... but I guess manual assembler use was out of fashion by the time so it wasn't added. It's just an alias really; even cmova is bitwise identical to cmovnbe.